README.md
# Qtlizer: comprehensive QTL annotation of GWAS results
[![Twitter](https://img.shields.io/twitter/url/http/shields.io.svg?style=social)](https://twitter.com/intent/tweet?hashtags=Qtlizer&url=https://www.biorxiv.org/content/10.1101/495903v2)


&nbsp;+ [Introduction](#Introduction)\
&nbsp;+ [Installation](#Installation)\
&nbsp;+ [Usage](#Usage)\
&nbsp;&nbsp;&nbsp;&nbsp;|-- [Accepted query terms](#Accepted-query-terms)\
&nbsp;&nbsp;&nbsp;&nbsp;|-- [Optional parameters](#Optional-parameters)\
&nbsp;&nbsp;&nbsp;&nbsp;|-- [Meta information](#Meta-information)\
&nbsp;+ [Try out online](#Try-out-online)\
&nbsp;+ [Authors](#Authors)\
&nbsp;+ [Citation](#Citation)\
&nbsp;+ [License](#License)


## Introduction
This **R** package provides access to the **Qtlizer** web server. **Qtlizer** annotates lists of common small variants (mainly SNPs) and genes in humans with associated changes in gene expression using the most comprehensive database of published quantitative trait loci (QTLs).

Alternatively, the **Qtlizer** can be accessed by using a web-based GUI (http://genehopper.de/qtlizer). More information about usage and available datasets can be found at http://genehopper.de/help#qtlizer_docu.


## Installation
```R
devtools::install_github('matmu/Qtlizer', build_vignettes = TRUE)
```

**Please note**: A valid internet connection (HTTP port: 80) is required in order to install and use the package.


## Help pages
```R
browseVignettes(Qtlizer)
help(package="Qtlizer")
```


## Usage
Simply call the function `get_qtls()` function to make requests to **Qtilzer**. The function utilizes a REST API (http://genehopper.de/rest) to query the annotation database. The QTL results will be returned as data frame or as `GenomicRanges::GRanges` object.

```R
get_qtls('rs4284742 DEFA1')
```
Common seperators (space, comma, space + comma, ...) are accepted. It is also possible to pass your query with a vector: 

```R
get_qtls(c("rs4284742", "DEFA1"))
```


### Accepted query terms
Accepted query terms are variant and gene identifiers of the form: 

+ Rsid : rs + number e.g. "rs4284742"
+ reference:chr:pos e.g. "hg19:19:45412079" (Allowed references: hg19/GRCh37, hg38/GRCh38; accepted chromosomes are 1-22)
+ Gene symbol consisting of letters and numbers according to  [https://www.genenames.org/about/guidelines/]( https://www.genenames.org/about/guidelines/)


### Optional parameters
+ corr: Correlation threshold based on linkage disequilibrium (LD) calculated from the 1000 Genomes Phase 3 European dataset. Optional value between 0 and 1. Default value is NA.
+ ld_method: Method to calculate correlation. Valid values are either "r2" (default) or "dprime".

	```R
	get_qtls("rs4284742", corr = 0.6, ld_method="r2")
	```

+ return_obj: Determinse the format of the result. Value "dataframe" (default) returns a dara frame whereas "grange" returns a ```GenomicRanges::GRanges``` object. 
+ ref_version: If output is a GRange object, the version of the reference genome is also considered. Accepted reference genome versions are "hg19" (default) or "hg38".

	```R
	get_qtls("rs4284742", return_obj = "grange", ref_version = "hg38")
	```

+ max_terms: Number of queries made at a time. The default value is 5. It is recommended to not set the value higher than 5. 

	```R
	get_qtls("rs4284742", max_terms = 4)
	```

### Meta information
Column descriptions of the received data frame can be accessed by calling:

```R
df = get_qtls("rs4284742")
comment(df)
```

## Try out online
If you want to try out the R package online, there is an example **Google Colaboratory** project at

https://colab.research.google.com/drive/1i1sjQHCjaw2wYzVBnXQ9iaghnk-jSU95#scrollTo=5Hi6sCe7SPFb

To run the project, make a private copy or open the project in playground mode and sign in to Google. 


## Authors
Matthias Munz [![](https://img.shields.io/twitter/follow/_MatthiasMunz?label=Follow&style=social)](https://img.shields.io/twitter/follow/_MatthiasMunz?label=Follow&style=social)\
Julia Remes\
University of Lübeck, Germany


## Citation
Please cite the following article when using **Qtlizer**:

**Munz M**, Wohlers I, Simon E, Busch H, Schaefer A<sup>\*</sup> and Erdmann J <sup>\*</sup> (2018) Qtlizer: comprehensive QTL annotation of GWAS results. ***bioRxiv***

[![](https://img.shields.io/badge/doi-https%3A%2F%2Fdoi.org%2F10.1101%2F495903%20-green.svg)](https://doi.org/10.1101/495903)
[![](https://img.shields.io/badge/Altmetric-17-green.svg)](https://www.altmetric.com/details/52777590)


## License
GNU General Public License v3.0



NEWS
Changes in version 1.8.1 (2022-02-28)
+ Citation adapted

Changes in version 0.1.2 (2020-04-21)
+ Minor changes

Changes in version 0.99.0 (2019-08-12)
+ Submitted to Bioconductor

Changes in version 0.99.4 (2019-08-14)
+ Added parameter ld_method and corr

Changes in version 0.99.5 (2019-08-27)
+ Added parameter return_obj to have the possibility to get a GenomicRanges::GRanges object

Changes in version 0.99.6 (2019-08-27)
+ Code revised

Changes in version 0.99.7 (2019-08-27)
+ Minor

Changes in version 0.99.8 (2019-08-27)
+ Additional HTTP error handling

Changes in version 0.99.9 (2019-08-27)
+ Bugfixes, error handling

Changes in version 0.99.10 (2019-08-27)
+ Minor

Changes in version 0.99.11 (2019-09-03)
+ Incorporated Bioconductor reviewer comments

Changes in version 0.99.12 (2019-09-06)
+ eval=false for package installation in vignette 
+ Title change in vignette


R/qtFunction.R
#'Query Qtlizer
#'
#'@description Makes a query at Qtlizer and writes the results 
#'into a specified (or default) file.
#'
#'@param query The qtlizer query.
#'@param outputFile (optional) The name of the output file.
#'@return Response goes directly into the output file.
#'@examples qtfun("rs4284742")
#'qtfun("rs4284742", "outputFile.txt")
#'@importFrom utils write.table
#'@export
qtfun <- function(query, outputFile=of_default){
    #Note: in the current version of Qtlizer, queries including 
    #gene identifier and chromosomal positions
    #of variants taker longer than if using rsids only

    of_default <- "defaultFile.txt"
    output_file <- outputFile
    q <- query
    corr <- 0.8 # optional
    ld_method <- "r2" # optional
    url <- paste('http://genehopper.de/rest/qtlizer?q=', gsub("\\s+", ",", q), 
        "&corr=", corr, "&ld_method=", ld_method, sep="")
    response <- httr::POST(url)
    result <- httr::content(response)

    a <- unlist(strsplit(result , "\n"))
    meta <- grep("^#", a, value = TRUE)
    data <- grep("^[^#]", a, value = TRUE)
    header <- unlist(strsplit(data[1], "\t"))
    ncols <- length(header)

    if(is.null(data[-1])){
        data <- NA
    } else {
        data <- unlist(strsplit(data[-1], "\t"))
    }

    m <- matrix(data, ncol=ncols, byrow=TRUE)
    d <- as.data.frame(m, stringsAsFactors=FALSE)

    write.table(meta, file=output_file, quote = FALSE, append=FALSE,
        row.names = FALSE, col.names = FALSE)

    write.table(gsub(", ", "\t",toString(header)), file=output_file, 
        quote = FALSE,  append=TRUE, 
        row.names = FALSE, col.names = FALSE)

    write.table(d, file=output_file, sep="\t", quote = FALSE, 
        append=TRUE, row.names = FALSE, col.names = FALSE)

}


R/get_qtls.R
#'Query Qtlizer
#'
#'@description Query Qtlizer database for expression quantitative 
#'trait loci (eQTLs) in human.
#'
#'@param query The query consists of search terms and can be a single 
#'string or a vector. Qtlizer allows to query both variants 
#'(Rsid, ref_version:chr:pos) and genes (Symbol consisting of letters 
#'and numbers according to the HGNC guidelines). The minimum allowed character length per search term is 2.
#'@param corr Linkage disequilibrium based on 1000 Genomes Phase 3 European. 
#'If this optional value between 0 and 1 is set, 
#'the input variants are enriched for proxy variants passing the threshold. 
#'Default value is NA. 
#'@param max_terms Number of terms in a single HTTP request. Default value is 5. 
#'A large value can lead to a very large result set and a error by the database.
#'@param ld_method There are two methods available: "r2" (default) and "dprime".
#'@param ref_version Two possible versions are supported: hg19 (GRCh37) or 
#'hg38 (GRCh38). Default value is "hg19". 
#'This argument is only considered if a GenomicRanges::GRanges object is returned.
#'@param return_obj The user can choose to get the QTL data to be returned 
#'as data frame or as a GenomicRanges::GRanges object. The default value is "dataframe".
#'@return Data frame or GenomicRanges::GRanges object containing QTL data.
#'@examples get_qtls("rs4284742")
#'get_qtls(c("rs4284742", "DEFA1"))
#'get_qtls("rs4284742,DEFA1")
#'get_qtls("rs4284742", return_obj="granges", ref_version="hg38")
#'get_qtls("rs4284742", corr=0.6)
#'@export
get_qtls = function(query, corr = NA, max_terms = 5, ld_method = "r2", 
                    ref_version = "hg19", return_obj = "dataframe"){
  
  # Check if there is an internet connection
  if (!curl::has_internet())
    stop("No internet connection detected...")
  
  
  # Split terms
  query = paste(query, sep = ' ', collapse = ' ')
  query = unique(unlist(strsplit(query, "[\\n\\s,\\t.;]+", perl=TRUE)))
  query = query[stringi::stri_length(query)>1]
  message(length(query)," unique query term(s) found")
  
  
  # Binningtolower
  bins = vector_split(query, ceiling(length(query)/max_terms))
  bins = stringi::stri_join_list(bins, sep = ",")
  
  
  # Communicate with database
  message("Retrieving QTL information from Qtlizer...")
  res = lapply(bins, communicate, corr = corr, ld_method = ld_method)
  res = do.call(rbind, res)
  
  if(!is.null(res)){
    message(nrow(res), " data point(s) received")
  } else {
    return(res)
  }
  
  
  # Convert to respective data types
  res[res == "-"] = NA
  if(all(c("p", "distance", "n_qtls", "n_best", "n_sw_sign", "n_occ") %in% names(res))){
    res$distance = as.numeric(res$distance)
    res$p = as.numeric(res$p) 
    res$n_qtls = as.numeric(res$n_qtls)
    res$n_best = as.numeric(res$n_best) 
    res$n_sw_sign = as.numeric(res$n_sw_sign)
    res$n_occ = as.numeric(res$n_occ)
  }
  
  
  # Create GRanges container
  if(tolower(return_obj) == "granges"){
    
    if(tolower(ref_version) == "hg19" || tolower(ref_version) == "grch37"){
      
      # Check for missing hg19 positions
      res_with_hg19 = res[which(!is.na(res$var_pos_hg19)),] 
      if(nrow(res_with_hg19) < nrow(res)){
        message("Not all results in GRanges object included due to missing hg19 (GRCh37) positions. Please use set ref_version to hg38 (GRCh38) andd/or set return_obj to 'dataframe' to obtain all results.")
      }
      
      gres = GenomicRanges::makeGRangesFromDataFrame(res_with_hg19, start.field = "var_pos_hg19", end.field = "var_pos_hg19", 
                                                     seqnames.field = "chr", keep.extra.columns = TRUE)
      
    } else {
      gres = GenomicRanges::makeGRangesFromDataFrame(res, start.field = "var_pos_hg38", end.field = "var_pos_hg38", 
                                                     seqnames.field = "chr", keep.extra.columns = TRUE)
    }
    
    GenomicRanges::strand(gres) = "+" 
    comment(gres) = comment(res)
    
    return(gres)
    
  } else {
    return(res)
  }
}


#'URL building and request/response handling
#'@param q The qtlizer query. Can either be a single string or a vector.
#'@param corr Linkage disequilibrium based on 1000 Genomes Phase 3 European.
#' Optional value between 0 and 1. Default value is NA. 
#'@param ld_method There are two methods. Default method is "r2". 
#'The other opportunity is to use "dprime".
#'@return Data frame with results.
#'@keywords internal
communicate = function(q, corr, ld_method, n.tries=2){
  
  
  # Build URL
  url = paste0('http://genehopper.de/rest/qtlizer?q=', q)
  
  if(is.numeric(corr) && corr>=0 && corr<=1){
    url = paste0(url, "&corr=", round(corr, digits=2))
    
    if(is.character(ld_method) && (ld_method == "r2" || ld_method == "dprime")){
      url = paste0(url, "&ld_method=", ld_method)
    }
  }
  
  
  # Send POST request and retrieve response
  while(n.tries > 0){
    response = httr::POST(url)
    
    if(!httr::http_error(response)){
      break
    }
    
    Sys.sleep(3)
    n.tries = n.tries-1
  }
  
  
  if(n.tries == 0){
    stop("Web server seems to be down. Try again later!")
  }
  
  
  result = httr::content(response)
  result = unlist(strsplit(result , "\n"))
  
  
  # Display error message from server
  if(length(grep("^#", result)) == 0){
    warning(result)
    return()
  }
  
  
  # Extract
  meta = grep("^#", result, value = TRUE)
  data = grep("^[^#]", result, value = TRUE)
  
  
  # If no QTLs were found
  if(length(data)-1 <= 0) {
    message("No QTLs found")
    return()
  }
  
  
  # Convert to dataframe
  header = unlist(strsplit(data[1], "\t"))
  ncols = length(header)
  data = unlist(strsplit(data[-1], "\t"))
  
  m = matrix(data, ncol=ncols, byrow=TRUE)
  d = as.data.frame(m, stringsAsFactors=FALSE)
  colnames(d) = header
  comment(d) = meta
  
  return(d)
}


#'Splits vector v into n subvectors
#'@param v input vector
#'@param n number of subvectors
#'@return List with subvectors.
#'@keywords internal
vector_split = function(v, n) {
  l = length(v)
  r = l/n
  return(lapply(seq_len(n), function(i){
    s = max(1, round(r*(i-1))+1)
    e = min(l, round(r*i))
    return(v[s:e])
  }))
}
